test/fenv/fenv_gentests.cc

```diff
@@ -0,0 +1,280 @@
+  fenv_t fenv;
+  std::fesetround( rounding );
+  std::feholdexcept( &fenv );
+
+  volatile T result  = func( ops... );
+  int        excepts = std::fetestexcept( FE_ALL_EXCEPT );
+
+  out << "  []{\n";
+  out << "    // Test " << testnum << "\n";
+  out << "    using namespace std;\n";
+  out << "    auto func = " << func_src << ";\n";
+  out << "    auto func_src = R\"(" << func_src << "\n)\";\n";
+  out << "    fenv_t fenv;\n";
+
+  switch( rounding ) {
+  case FE_TONEAREST: out << "    fesetround( FE_TONEAREST );\n"; break;
+  case FE_TOWARDZERO: out << "    fesetround( FE_TOWARDZERO );\n"; break;
+  case FE_UPWARD: out << "    fesetround( FE_UPWARD );\n"; break;
+  case FE_DOWNWARD: out << "    fesetround( FE_DOWNWARD );\n"; break;
+  }
+
+  out << "    feholdexcept( &fenv );\n";
+  out << "    " << type<T> << " result{ func( " << args_string( ops... ) << " ) };\n";
+  out << "    int exceptions = fetestexcept( FE_ALL_EXCEPT );\n";
+  out << "    " << type<T> << " result_expected{ " << repr( result ) << " };\n";
+  out << "    int exceptions_expected = " << exception_string( excepts ) << ";\n";
+  out << "    bool result_passed = test_result( \"" << testnum << "\", file_prefix, func_src, result, result_expected, "
+      << args_string( ops... ) << " );\n";
+  out << "    bool exceptions_passed = test_exceptions( \"" << testnum
+      << "\", file_prefix, func_src, exceptions, exceptions_expected, result_passed, " << args_string( ops... ) << ");\n";
+  out << "    failures += !(result_passed && exceptions_passed);\n";
+  out << "    fesetenv( &fenv );\n";
+  out << "  },\n";
+  std::fesetenv( &fenv );
+
+  if( ++testcount >= maxtests_per_file ) {
+    closefile();
+    testcount = 0;
+  }
+}
+
+template<typename FP>
+static const FP special_fp_values[] = {
+  0.0f,
+  -0.0f,
+  1.0f,
+  -1.0f,
+  1.5f,
+  -1.5f,
+  std::numeric_limits<FP>::quiet_NaN(),
+  std::numeric_limits<FP>::signaling_NaN(),
+  std::numeric_limits<FP>::infinity(),
+  -std::numeric_limits<FP>::infinity(),
+  std::numeric_limits<FP>::lowest(),
+  std::numeric_limits<FP>::max(),
+};
+
+template<typename FP, typename INT>
+static const FP special_fcvt_values[] = {
+  FP( std::numeric_limits<INT>::max() ),
+  FP( std::numeric_limits<INT>::max() ) + FP( 0.5 ),
+  FP( std::numeric_limits<INT>::max() ) + FP( 1.0 ),
+  FP( std::numeric_limits<INT>::max() ) - FP( 0.5 ),
+  FP( std::numeric_limits<INT>::max() ) - FP( 1.0 ),
+  FP( std::numeric_limits<INT>::min() ),
+  FP( std::numeric_limits<INT>::min() ) + FP( 0.5 ),
+  FP( std::numeric_limits<INT>::min() ) + FP( 1.0 ),
+  FP( std::numeric_limits<INT>::min() ) - FP( 0.5 ),
+  FP( std::numeric_limits<INT>::min() ) - FP( 1.0 ),
+  FP( 0 ),
+  -FP( 0 ),
+  FP( 0.1 ),
+  FP( -0.1 ),
+  FP( 0.5 ),
+  FP( -0.5 ),
+  FP( 0.9 ),
+  FP( -0.9 ),
+  std::numeric_limits<FP>::quiet_NaN(),
+  std::numeric_limits<FP>::signaling_NaN(),
+  std::numeric_limits<FP>::infinity(),
+  -std::numeric_limits<FP>::infinity(),
+};
```

Comment:
These are special FP values for when converting from FP to integer, with controlled rounding.